File: fibo/fibo.py
#!/bin/python27

def main():
   print fiboD(38)
   print fibo(38)

# Dumb fibo recursion
def fibo(n):
   '''
   Given an index returns that value of the fibonacci sequence. Using dumb
   recursion.
   >>> fibo(7)
   13
   '''
   # This assumes n is positive
   if(n<2):
      return n # return 1 or 0
   return fibo(n-1) + fibo(n-2)

# Container function for Dynamic recursion
def fiboD(n):
   '''
   Removes extra return value produce by the dynamic recursion function (_fiboD)
   >>> fiboD(7)
   13
   '''
   a,b=_fiboD(n)
   return a

# Actual Dynamic recursion function, note the parameters act as return varibles
def _fiboD(n):
   '''
   Given an index returns that value of the fibonacci sequence and it's 
   previous value. Using dynamic recursion.
   >>> _fiboD(7)
   (13, 8)
   '''
   # This assumes n is positive
   if(n<2):
      return n,0 # returns 1,0 or 0,0
   a,b = _fiboD(n-1)
   return a+b, a
   
if __name__=='__main__':
   main()

File: fibo/fibo.c
#include <stdio.h>

int fibo(int);
int fiboD(int);
void _fiboD(int*, int*);

main()
{
  printf("%i\n", fiboD(38));
  printf("%i\n", fibo(38));
}

/* Dumb fibo recursion */
int fibo(int n)
{
   /* This assumes n is positive  */
   if(n<2)
      return n; // Returns 1 or 0  */
   return fibo(n-1) + fibo(n-2);
}

/* Container function for Dynamic recursion */
int fiboD(int n)
{
  int b;
  _fiboD(&n, &b);
  return n;
}

/* Actual Dynamic recursion function, note the parameters act as return varibles */
void _fiboD(int *a, int *b)
{
  if(*a<2) {
    *b = 0;
    return; /* a=1,b=0 or a=0,b=0 */
  }
  --*a; 
  _fiboD(a,b);
  
  *a += *b; 
  *b = *a - *b;
  return; /* a=a+b, b=a */
}
